File: src/data/types/resourceInstance.ts
import type { Resource } from "./resource.ts";
import type { QualityRoll } from "./quality.ts";

// A concrete, rolled batch of a Resource -- what refine()/craft() actually
// consume. Multiple instances of the same Resource type can carry different
// rolled qualities (different gathering rolls), so quality lives per
// instance, not per Resource.
export interface ResourceInstance {
  resource: Resource;
  quantity: number;
  qualities: QualityRoll;
}

File: unity/ProfitableCore/Schema/ResourceInstance.cs
namespace Profitable.Core.Schema;

// Ports src/data/types/resourceInstance.ts. Necessary completion (see
// agent-32-unity-simulation-core.md's Outputs Section 1) -- a concrete,
// rolled batch of a Resource, what Refine()/Craft() actually consume.
// Multiple instances of the same Resource can carry different rolled
// qualities (different gathering rolls), so quality lives per instance,
// not per Resource.
public class ResourceInstance
{
    public Resource Resource { get; set; } = new();
    public int Quantity { get; set; }
    public QualityMap Qualities { get; set; } = new();
}
